src/bin/pg_autoctl/supervisor.c

```diff
@@ -743,43 +741,31 @@ supervisor_may_restart(Service *service)
 			  epoch_to_string(counters->startTime[position], timestring),
 			  (int) (now - counters->startTime[position]));
 
-	for (; position != counters->position || count == 0;)
+	/* until we have restarted MaxR+1 times, we know we can restart */
+	if (counters->count <= SUPERVISOR_SERVICE_MAX_RETRY)
 	{
-		uint64_t restartTime = counters->startTime[position];
-
-		/* if we have ever restarted once, only loop once */
-		if (count == counters->count)
-		{
-			break;
-		}
-
-		if ((now - restartTime) <= SUPERVISOR_SERVICE_MAX_TIME)
-		{
-			/* we loop through the ring buffer in restart order */
-			oldestRestart = restartTime;
-			restarts++;
-
-			log_trace("supervisor_may_restart: service \"%s\" restarted "
-					  " at %s, %d seconds ago",
-					  service->name,
-					  epoch_to_string(restartTime, timestring),
-					  (int) (now - restartTime));
-		}
-
-		count++;
-
-		/* progress backward in time in the ring buffer */
-		position = position == 0 ? size - 1 : position - 1;
+		return true;
 	}
 
-	if (restarts > SUPERVISOR_SERVICE_MAX_RETRY)
+	/*
+	 * When we have restarted MaxR+1 times or more, the only case when we can't
+	 * restart again is if the oldest entry in the counters startTime array is
+	 * older than our MaxT.
+	 *
+	 * The oldest entry in the ring buffer is the one just after the current
+	 * one:
+	 */
+	position = (position + 1) % size;
+	oldestRestartTime = counters->startTime[position];
+
+	if ((now - oldestRestartTime) <= SUPERVISOR_SERVICE_MAX_TIME)
 	{
 		log_fatal("pg_autoctl service %s has already been "
 				  "restarted %d times in the last %d seconds, "
 				  "stopping now",
 				  service->name,
-				  restarts,
-				  (int) (now - oldestRestart));
+				  size,
```

Comment:
I believe it should be this:
```suggestion
				  SUPERVISOR_SERVICE_MAX_TIME,
```

Reply:
I don't think so, because we allow up to MaxR restarts. We fail at MaxR+1, which is why the size of the ring buffer has to be MaxR+1 in the first place.

From https://erlang.org/doc/design_principles/sup_princ.html

> If more than MaxR number of restarts occur in the last MaxT seconds, the supervisor terminates all the child processes and then itself. The termination reason for the supervisor itself in that case will be shutdown.

Reply:
I don't think we need ring buffer size of MaxR+1. We only store the current restart after checking the oldest. So we could overwrite that. Then we would still restart 5 times always and at the 6th time we consider if the entry we are about to overwrite is in our MaxT. If it is, we don't restart, because that would be restart MaxR+1 in MaxT seconds. If it is not, we can throw it out, because we do not need it anymore.

Reply:
Ah, yeah. Done that way now.